pennylane/clusters.py

```diff
@@ -0,0 +1,476 @@
+        template (callable): the ansatz for the circuit before the final measurement step
+        observables (Iterable[:class:`~.Observable`]): observables to measure during the
+            final step of each circuit
+        device (Device, List[Device]): Corresponding device(s) where the resulting
+            QNodeCluster should be executed. This can either be a single device, or a list
+            of devices of length ``len(observables)``.
+        measure (str, List[str]): Measurement(s) to perform. Options include
+            :func:`'expval' <pennylane.expval>`, :func:`'var' <pennylane.var>`,
+            and :func:`'sample' <pennylane.sample>`.
+            This can either be a single measurement type, in which case it is applied
+            to all observables, or a list of measurements of length ``len(observables)``.
+        interface (str, None): which interface to use for the returned QNode cluster.
+            This affects the types of objects that can be passed to/returned from the QNode:
+
+            * ``interface='autograd'``: Allows autograd to backpropogate
+              through the QNodeCluster. The QNodeCluster accepts default Python types
+              (floats, ints, lists) as well as NumPy array arguments,
+              and returns NumPy arrays.
+
+            * ``interface='torch'``: Allows PyTorch to backpropogate
+              through the QNodeCluster.The QNodeCluster accepts and returns Torch tensors.
+
+            * ``interface='tf'``: Allows TensorFlow in eager mode to backpropogate
+              through the QNodeCluster. The QNodeCluster accepts and returns
+              TensorFlow ``tf.Variable`` and ``tf.tensor`` objects.
+
+            * ``None``: The QNodeCluster accepts default Python types
+              (floats, ints, lists) as well as NumPy array arguments,
+              and returns NumPy arrays. It does not connect to any
+              machine learning library automatically for backpropagation.
+
+        diff_method (str, None): the method of differentiation to use in the created QNodeCluster.
+
+            * ``"best"``: Best available method. Uses the device directly to compute
+              the gradient if supported, otherwise will use the analytic parameter-shift
+              rule where possible with finite-difference as a fallback.
+
+            * ``"parameter-shift"``: Use the analytic parameter-shift
+              rule where possible with finite-difference as a fallback.
+
+            * ``"finite-diff"``: Uses numerical finite-differences.
+
+            * ``None``: a non-differentiable QNodeCluster is returned.
+
+    Returns:
+        QNodeCluster: a cluster of QNodes executing the circuit template with
+        the specified measurements
+
+    **Example:**
+
+    Let's define a custom two-wire template:
+
+    .. code-block:: python
+
+        def my_template(params, wires, **kwargs):
+            for i in range(2):
+                qml.RX(params[i], wires=wires[i])
```

Comment:
This indicates that the circuit is dynamic in terms of the number of parameters passed, whereas:
*  If ``len(params)==2``, then it might be better to explicitly have 
    ```
    qml.RX(params[0], wires=wires[0])
    qml.RX(params[1], wires=wires[1])
    ```
*  If ``len(params)`` is dynamically changing, then having ``for i in range(len(params)):`` as the loop would be better to have

Reply:
Good idea!